rancher/rancher
# Pin CAPI chart version

## Issue: https://github.com/rancher/rancher/issues/47848
 
## Problem
The version of the CAPI chart that is automatically installed in downstream clusters is no longer being pinned in the released rancher images after the switch to GHA.
 
## Solution
Pin the chart version by passing the value as a build argument to the rancher docker image, both server and agent, in the same way this is done for other charts (e.g. the Webhook chart).

This PR also makes related fixes:
- It changes the provisioning test script (`scripts/provisioning-tests`) and the local integration tests (in `scripts/test`) to take the pinned CAPI chart version.
- It passes the chart version argument when building the server image in `dev-scripts/quick`.

The provisioning tests run through GHA use the rancher binary directly, and not the docker image previously built in the CI, so explicitly passing the chart version in the script is necessary. Running the provisioning tests locally will also take the pinned version now.

The integration tests already use docker image, so no changes are needed to `scripts/gha/tests`. The local integration tests (run by `script/test`) now also pass the chart version to the rancher instance that it runs.

## Additional points

I considered baking the CAPI version into the rancher binary as a default to the corresponding setting, instead of using an environment variable for the setting, because it's available from `pkg/buildconfig/constants.go`, which is generated from `build.yaml`, the source of truth the version. The benefit would be to not worry about all the places where the environment has to be set (e.g. tests, docker images). However, there is a difference in behavior between setting the variable through the environment and setting a default: values from the environment are always picked over the actual value of the setting object, so this prevents changes to the setting object in etcd to have any effect. Defaults can be overridden by altering the setting object. The chart version could be pinned directly, without going through `settings.go` and setting objects, however, this would be too different from every other setting and might cause confusion, so I preferred keeping this like all the other chart versions (e.g. webhook).

For the provisioning test script and the local integration test script, I opted to only export the capi version, instead of calling `scripts/export-config` to export all settings from `build.yaml` to avoid chaning other behavior. For instance, the `CATTLE_DEFAULT_SHELL_VERSION` setting is not set in the Docker image, but it's directly set as a default in `setting.go`. This would change how it's read in the tests.
 
## Testing


## Engineering Testing
### Manual Testing
- For the `quick` script: inspected the generated images and checked that the environment variable was correctly set.
- For the local tests (integration and provisioning): temporarily modified rancher to print the setting corresponding to the capi chart version and dumped the logs at the end of the tests. Checked that the version was set.
- For the provisioning and integration test GH actions: temporarily dumped the logs and printed the chart version, checked that it was set.
- Checked in the action logs that the capi version is being passed as a build arg to the server and agent.
- Downloaded and loaded the image artifacts built by the actions to check if the environment variable is set.
- Use these images to run rancher and follow the reproduction steps, and verified that the chart was not updated from the pinned version.

### Automated Testing
* Test types added/modified:
    * None
* If "None" - Reason: This changes how environment variables are passed rancher, not rancher itself.

## QA Testing Considerations
 
### Regressions Considerations
There is no newer version of this chart on this branch, so the selected version will be the same as now. This only prevents the chart from being upgraded on future new versions.

Existing / newly added automated tests that provide evidence there are no regressions:
* All the provisioning tests
